### PR TITLE
Process Safety

### PR DESCRIPTION
**Pull request by zovt**
### Proposed changes in this pull request
Make Phantom class more process-safe

See #495 

#### Checklist
* [x] New tests have been added
* [x] `npm test` passes successfully
* [ ] Documentation has been updated

@amir20 to review

### CODE REVIEW

File: src/phantom.js [hunk 1]
```diff
@@ -219,12 +224,32 @@ export default class Phantom {
...
     _heartBeat() {
         if (this.commands.size === 0) {
             this.execute('phantom', 'noop');
         }
     }
+
+    /**
+     * rejects all commands in this.commands
+     */
+    rejectCommands(errmsg = 'Phantom exited prematurely') {
```
Comment (amir20) [inline src/phantom.js]: Can you make all the methods that shouldn't be used externally to start with `_` so like `_rejectCommands`

File: src/phantom.js [hunk 2]
```diff
@@ -75,9 +76,15 @@ export default class Phantom {
         this.process.on('error', error => {
             logger.error(`Could not spawn [${phantomjs.path}] executable. Please make sure phantomjs is installed correctly.`);
             logger.error(error);
+            this.kill('Process got an error');
             process.exit(1);
         });
 
+        this.process.stdin.on('error', (e) => {
```
Comment (amir20) [inline src/phantom.js]: Is there a way to add a test for this?

File: src/phantom.js [hunk 3]
```diff
@@ -75,9 +76,15 @@ export default class Phantom {
...
+            logger.debug(`Child process received error ${e}, sending kill signal`);
+            this.kill('Error reading from stdin');
```
Comment (amir20) [inline src/phantom.js]: Is this really what you want here? If any stdin error is outputted then the process just exits. That doesn't sound right. What if PhantomJS does prints a warning or error but everything else still works.

**Comment by amir20** [comment]
Tests are missing 👎

File: src/command.js [hunk 4]
```diff
@@ -11,4 +11,4 @@ export default class Command {
         this.params = params;
         this.deferred = undefined;
     }
-}
\ No newline at end of file
+}
```
Comment (amir20) [inline src/command.js]: What's with all the new lines? Were they added or removed?

File: src/phantom.js [hunk 5]
```diff
@@ -219,12 +224,32 @@ export default class Phantom {
...
     exit() {
         clearInterval(this.heartBeatId);
-        this.execute('phantom', 'invokeMethod', ['exit']);
+        return this.execute('phantom', 'invokeMethod', ['exit']);
+    }
+
+    /**
+     * Clean up and force kill this process
+     */
+    kill(errmsg = 'Phantom process was killed') {
```
Comment (amir20) [inline src/phantom.js]: Should be `_kill` since you probably don't want people calling it externally

File: src/phantom.js (hunk 1 above)
Comment (amir20) [inline src/phantom.js]: `_rejectAllCommands`? Since it is rejecting all commands

**Comment by zovt** [comment]
Hm. I'm pretty sure that `this.process.stdin.on('error')` is only triggered when there's an error reading or writing from the Stream, not when phantom prints a warning or error. I'll add the _s where necessary

File: src/phantom.js (hunk 5 above)
Comment (zovt) [inline src/phantom.js]: I think being able to kill the process externally could be ok

File: src/phantom.js [hunk 6]
```diff
@@ -75,9 +76,15 @@ export default class Phantom {
         this.process.on('error', error => {
             logger.error(`Could not spawn [${phantomjs.path}] executable. Please make sure phantomjs is installed correctly.`);
             logger.error(error);
+            this.kill('Process got an error');
```
Comment (amir20) [inline src/phantom.js]: I think this message should include `error` as well.

File: src/phantom.js (hunk 3 above)
Comment (amir20) [inline src/phantom.js]: Same thing here regarding `e` for message.

Comment (zovt) [inline src/phantom.js]: Hm. I'm pretty sure that this.process.stdin.on('error') is only triggered when there's an error reading or writing from the Stream, not when phantom prints a warning or error. I'll add the _s where necessary

File: src/command.js (hunk 4 above)
Comment (amir20) [inline src/command.js]: Nevermind, they were added. 👍

File: src/phantom.js (hunk 6 above)
Comment (zovt) [inline src/phantom.js]: The _rejectAllCommands method rejects with a new error, so when printed you get Error: errmsg

**Comment by amir20** [comment]
Overall it looks good. But a couple of things: 

1. Tests missing
2. private methods to start with `_*`
3. I am not sure if you'd want stdin.error to just kill every process. That sounds a little aggressive.

File: src/command.js (hunk 4 above)
Comment (zovt) [inline src/command.js]: Dunno, probably a quirk with emacs.

**Comment by zovt** [comment]
3. stdin.error should only kill the Phantom instance associated with the class
1. I'll try to create a test for this

File: src/phantom.js (hunk 3 above)
Comment (amir20) [inline src/phantom.js]: You may be right. I am going to google it.

File: src/phantom.js (hunk 5 above)
Comment (elias-winberg) [inline src/phantom.js]: It doesn't make sense to throw an error if the user deliberately/externally kills the process.

File: src/phantom.js (hunk 3 above)
Comment (amir20) [inline src/phantom.js]: I think you are right. `stderr` will get used instead which I handled correctly. 👍

File: src/phantom.js (hunk 6 above)
Comment (amir20) [inline src/phantom.js]: The message goes `kill` -> `rejectAllCommands` -> `command.deferred.reject` so how does it create a new error? I think it will just end up being `new Error("Process got an error");` right? 

This message should probably something more clear. Like `this.kill('Could not spawn [${phantomjs.path}] executable')`

File: src/phantom.js [hunk 7]
```diff
@@ -219,12 +224,32 @@ export default class Phantom {
...
+        return this.execute('phantom', 'invokeMethod', ['exit']);
+    }
+
+    /**
+     * Clean up and force kill this process
+     */
+    kill(errmsg = 'Phantom process was killed') {
+        this.rejectCommands(errmsg);
+        this.process.kill('SIGKILL');
+        return Promise.reject(errmsg);
```
Comment (amir20) [inline src/phantom.js]: Does this returned `Promise` actually do anything? I can't see to find where it matters.

**Comment by amir20** [comment]
Woohoo another person reviewing. Thanks @elias-winberg

File: src/phantom.js (hunk 7 above)
Comment (elias-winberg) [inline src/phantom.js]: No, since `phantom.exit()` is synchronous the process will have exited before any response is sent.

**Comment by zovt** [comment]
@elias-winberg Not sure. How would I test that?

File: src/phantom.js (hunk 7 above)
Comment (elias-winberg) [inline src/phantom.js]: The promise will be pending forever.

Comment (zovt) [inline src/phantom.js]: Just in case a user would want to wait for the exit method to finish 'synchronously'

**Comment by elias-winberg** [comment]
@zovt Never mind! I confused `stdin` with `stderr`.

File: src/phantom.js (hunk 7 above)
Comment (zovt) [inline src/phantom.js]: phantom.exit() meaning PhantomJs's phantom.exit() or the class's?

File: src/phantom.js [hunk 8]
```diff
@@ -75,26 +75,15 @@ export default class Phantom {
         this.process.on('exit', code => logger.debug(`Child exited with code {${code}}`));
         this.process.on('error', error => {
             logger.error(`Could not spawn [${phantomjs.path}] executable. Please make sure phantomjs is installed correctly.`);
-	        logger.error(error);
-	        console.log(error);
-            this.process.exit(1);
+            logger.error(error);
+            this.kill();
+            process.exit(1);
```
Comment (elias-winberg) [inline src/phantom.js]: Why did you add this?

File: src/phantom.js (hunk 5 above)
Comment (zovt) [inline src/phantom.js]: the error is thrown to the rejected promises

Comment (zovt) [inline src/phantom.js]: Oh I see what you're saying now. Yes I suppose that doesn't make sense

File: src/phantom.js (hunk 6 above)
Comment (zovt) [inline src/phantom.js]: Oof my reading comprehension has gone out the door in this thread. Sorry about that guys. Yes, that should most certainly include the error

File: src/phantom.js (hunk 8 above)
Comment (zovt) [inline src/phantom.js]: that's what it was originally if you look at this repo. I don't know why that was there honestly. Seems like it shouldn't have been.

File: src/phantom.js (hunk 7 above)
Comment (elias-winberg) [inline src/phantom.js]: The promise will never be resolved, since no response is sent.

**Comment by zovt** [comment]
Any ideas on how to add the tests for the stdin error handler?

File: src/phantom.js (hunk 7 above)
Comment (zovt) [inline src/phantom.js]: got it. So is there any way to check for a successful exit then?

**Comment by elias-winberg** [comment]
I'm not sure, sorry! Also, if we handle `stdin` errors, shouldn't we also handle `stdout` errors?

**Comment by zovt** [comment]
Probably!

**Comment by elias-winberg** [comment]
Will an error occur if you end a `Stream` and try to write to it? If so, you can do just that to test the `stdin` error handler.

**Comment by elias-winberg** [comment]
I just checked, and an error isn't emitted - it's thrown.

**Comment by zovt** [comment]
How's that look?

File: src/spec/phantom_spec.js [hunk 9]
```diff
@@ -51,4 +52,18 @@ describe('Phantom', () => {
     it('#create(true) to throw an exception', () => {
         expect(() => new Phantom(true)).toThrow();
     });
-});
\ No newline at end of file
+
+    it('catches errors when stdin closes unexpectedly', (done) => {
+        instance.process.stdin.end();
+        instance.createPage().catch(() => {
```
Comment (elias-winberg) [inline src/spec/phantom_spec.js]: Don't you think the test is a bit too generic?

**Comment by zovt** [comment]
Not sure how much more specific it can get than that. If you have any ideas let me know

File: src/phantom.js [hunk 10]
```diff
@@ -75,9 +76,20 @@ export default class Phantom {
...
+            this.kill(`Process got an error: ${error}`);
             process.exit(1);
         });
 
+        this.process.stdin.on('error', (e) => {
+            logger.debug(`Child process received error ${e}, sending kill signal`);
+            this.kill(`Error reading from stdin: ${e}`);
+        });
+
+        this.process.stdout.on('error', (e) => {
```
Comment (amir20) [inline src/phantom.js]: Looking good

File: src/spec/phantom_spec.js [hunk 11]
```diff
@@ -51,4 +52,26 @@ describe('Phantom', () => {
     it('#create(true) to throw an exception', () => {
         expect(() => new Phantom(true)).toThrow();
     });
-});
\ No newline at end of file
+
+    it('catches errors when stdin closes unexpectedly', (done) => {
```
Comment (amir20) [inline src/spec/phantom_spec.js]: Looks good

**Comment by amir20** [comment]
Looks good. Will merge later today. Will let it sit for a while.

**Comment by zovt** [comment]
Sounds good. Thanks for the quick responses!

**Comment by amir20** [comment]
Merged and releases.
